synapse/storage/databases/main/events_worker.py

```diff
@@ -1462,68 +1468,68 @@ async def have_seen_events(
             event_ids: events we are looking for
 
         Returns:
-            The set of events we have already seen.
+            The remaining set of events we haven't seen.
         """
+        set_tag(
+            SynapseTags.FUNC_ARG_PREFIX + "event_ids.length",
+            str(len(event_ids)),
+        )
 
         # @cachedList chomps lots of memory if you call it with a big list, so
         # we break it down. However, each batch requires its own index scan, so we make
         # the batches as big as possible.
 
-        results: Set[str] = set()
-        for chunk in batch_iter(event_ids, 500):
-            r = await self._have_seen_events_dict(
-                [(room_id, event_id) for event_id in chunk]
-            )
-            results.update(eid for ((_rid, eid), have_event) in r.items() if have_event)
+        remaining_event_ids: Set[str] = set()
+        for chunk in batch_iter(event_ids, 1000):
+            remaining_event_ids_from_chunk = await self._have_seen_events_dict(chunk)
+            remaining_event_ids.update(remaining_event_ids_from_chunk)
 
-        return results
+        return remaining_event_ids
 
-    @cachedList(cached_method_name="have_seen_event", list_name="keys")
```

Comment:
Is there any max size of a `@cachedList`? Seems like it grows forever

Reply:
It ought to be the same cache as `have_seen_event`, which has `max_entries=100000`

Reply:
*Some discussion in [backend internal](https://matrix.to/#/!SGNQGPGUwtcPBUotTL:matrix.org/$Jk6dY7gXaadOF-VMI5WnRcCDZ-Q2_g1T3eg4y3wdMtU?via=jki.re&via=matrix.org&via=element.io),*

Seems like it currently uses the default of `1000` from the [`DeferredCache`](https://github.com/matrix-org/synapse/blob/f9f03426de338ae1879e174f63adf698bbfc3a4b/synapse/util/caches/deferred_cache.py#L74).

And it's not one of the `per_cache_factors` that we configure for [`matrix.org`](https://github.com/matrix-org/matrix-ansible-private/blob/29880bdebaf99d898a5a659fcf6b73caf54e5c7c/inventory/group_vars/matrixorg_synapse/workers) so it's just some small multiple of that in any case with the `global_factor`.

Reply:
> It ought to be the same cache as have_seen_event, which has max_entries=100000 (before multiplying by the global cache factor).

@squahtx Ahhh, I think see what you're trying to say. `_have_seen_events_dict()` re-uses the `have_seen_event` cache which is defined at `max_entries=100000`

https://github.com/matrix-org/synapse/blob/f9f03426de338ae1879e174f63adf698bbfc3a4b/synapse/storage/databases/main/events_worker.py#L1481-L1482

https://github.com/matrix-org/synapse/blob/f9f03426de338ae1879e174f63adf698bbfc3a4b/synapse/storage/databases/main/events_worker.py#L1523-L1524

---

In any case, it seems to have a 0% cache hit rate,

https://grafana.matrix.org/d/000000012/synapse?orgId=1&from=1660891259284&to=1660934459284&viewPanel=1
![](https://user-images.githubusercontent.com/558581/185685728-4ef96905-911e-4f9e-b2ff-695042789578.png)

And the cache size doesn't get above 6k:

![](https://user-images.githubusercontent.com/558581/185685835-0cb3e0fb-a0d5-451c-98f5-e8980ea9d4a0.png)

Reply:
This cache still seems so weird to me. Why doesn't it grow bigger? We should at least see 200k state events from Matrix HQ show up and completely fill up the cache sometimes.

https://grafana.matrix.org/d/000000012/synapse?orgId=1&from=1664997921975&to=1666293921975&viewPanel=8
![](https://user-images.githubusercontent.com/558581/197039554-8d59c1c1-fd77-437f-99e8-98310ce3a5fd.png)